### PR TITLE
Add Payment Request button support for Woo blocks

### LINKED ISSUE #1416: Enable payment request buttons for checkout and cart blocks

### LINKED ISSUE #1544: Payment Request button + Block checkout compatibility with 3D Secure cards
Implement compatibility with Payment Request buttons + 3D Secure cards in the block checkout.

### LINKED ISSUE #1677: Payment Request Button - error when checkout phone field is not required and the product is virtual
To reproduce on Chrome:
- Enable payment request buttons
- Have a product that is virtual and downloadable (i.e. no shipping required)
- Go to `Appearance > Customize > WooCommerce > Checkout` and set the phone field to `Hidden`
- Attempt to pay for the product with the payment request button
- Note the dialog doesn't list the phone number or address:
Screenshot 2021-04-23 at 12 11 32

After submitting the payment dialog, an error appears in the console:
Screenshot 2021-04-23 at 12 10 45

### PR DESCRIPTION
**Pull request by ricardo**
Fixes #1416
Fixes #1544
Fixes #1677

#### Changes proposed in this Pull Request

- Add Payment Request blocks integration. The code is adapted from the Stripe Payment Gateway (1467-gh-woocommerce/woocommerce-gateway-stripe).
- Use same normalization functions and API in blocks and shortcode.
- Remove some leftover code from shortcode integration (e.g. `clearCart`, unnecessary i18n and nonce variables) and make shortcode integration more similar to the classic checkout by using the `WCPayAPI`.
- Fix 3D Secure / SCA confirmation with payment request buttons.

There are a few improvements left to do:
- Perhaps change folder structure, moving everything related to blocks into `client/blocks/`. - This requires changing the current credit card checkout as well.
- Add loading indicator to PRB when 3DS modal is open or intent is being validated. - Not really related to this PR, as it applies to the shortcode implementation too.

#### Testing instructions

- Run `npm run build:client`.
- Install Gutenberg and WooCommerce Blocks.
- Create a second cart page and add the cart block to it.
- Create a second checkout page and add the checkout block to it.
- Test using the payment request button in the product, cart and checkout pages, both shortcode and blocks.
- Verify that the button shows up as expected.
- Verify that the order is created as expected when the payment succeeds.

-------------------

- [x] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

**Comment by asumaran** [comment]
@ricardo I've tested my PR https://github.com/Automattic/woocommerce-payments/pull/1642 on top of your branch and didn't see any issue. My PR affects the Product page only.

**Comment by ricardo** [comment]
@asumaran Thanks for testing it. I was concerned the changes I made to the payment request button class could break your PR.

**Review by reykjalin — CHANGES_REQUESTED** [review]
Just noticed that Google Pay does not have a `light-outline` theme; only `dark` and `light`:

| PRB | Network request |
| :---: | :---: |
| image | image |

I think the fix for this is basically just to map `light-outline` to `light`?

**Review by reykjalin — CHANGES_REQUESTED** [review]
When you fail the 3DS authentication while paying with a PRB the error message isn't properly formatted:

image

image

**Review by reykjalin — CHANGES_REQUESTED** [review]
This is all coming along nicely! I think there's just some minor comments on the code, and the issues I pointed out ^ above.

I'm going to do some more testing, especially to make sure the shortcode flow is still working as expected following the changes to the API, but I think this is all generally looking pretty good ☺️ 

I really like how you've been able to adapt the Stripe implementation to fit better into the WCPay codebase. Excellent work! 👏

**Review by reykjalin — CHANGES_REQUESTED** [review]
**This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR**

When I tried to pay **through the credit card form** with a card that will be declined (`4000 0000 0000 0002`) I don't get the expected `Card declined` error, and instead get this generic message at the top of the checkout page:

| Error Message | Credit Card Form |
| :---: | :---: |
| image | image |

The PRB shows the expected error message in the correct place.

**Review by reykjalin — CHANGES_REQUESTED** [review]
It seems the [aforementioned issue with failing 3DS verification through a PRB](https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647223068) might be an issue that existed before this PR, since the product page shows the exact same result:

image

**Review by reykjalin — CHANGES_REQUESTED** [review]
Related to the [`light-outline` button not showing up correctly](https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647217822) it works on the Product page, so that's definitely new with this PR.

https://github.com/Automattic/woocommerce-payments/pull/1529#pullrequestreview-647217822

**Comment by ricardo** [comment]
> This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR

Please move to a new issue 👍 

I have seen this error before, but didn't connect the dots. Well observed though.

It's probably because the block credit card method is still using the generic Gutenberg checkout endpoint to create the order, and that's not throwing any specific response from Stripe.

File: client/payment-request/event-handlers.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	);
+
+	if ( 'success' === response.result ) {
+		try {
+			const confirmation = api.confirmIntent( response.redirect );
+			// We need to call `complete` outside of `completePayment` to close the dialog for 3DS.
+			event.complete( 'success' );
+
+			// `true` means there is no intent to confirm.
+			if ( true === confirmation ) {
```
[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/event-handlers.js:58]: This was taken from `confirm-card-payment.js` [here](https://github.com/Automattic/woocommerce-payments/blob/0ece4e0791971a08e622134ff60d0e417e083392/client/checkout/blocks/confirm-card-payment.js#L17), from the credit card checkout.

> Actually — after reading the comment — I'm thinking willNotConfirm might be a better name? Or intentWasRejected?

I didn't understand the reasoning behind calling it the opposite name. Can you explain?

It kinda makes sense to call it `confirmation` or `confirmed` because `api.confirmIntent` might return `true` at first, but then it can return a `confirmation` Promise, which then resolves to a redirect URL.

File: client/payment-request/utils/normalize.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			return cartTotalItem.value
+				? {
+						amount: cartTotalItem.value,
+						label: cartTotalItem.label,
+						pending,
+				  }
+				: false;
+		} )
+		.filter( Boolean );
+};
```
[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/utils/normalize.js]: Yes, makes sense 👍

File: client/payment-request/utils/utils.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ */
+export const getPaymentRequestData = ( key ) => {
+	if (
+		'object' === typeof wcpayPaymentRequestParams &&
+		wcpayPaymentRequestParams.hasOwnProperty( key )
+	) {
+		return wcpayPaymentRequestParams[ key ];
+	}
+	return null;
+};
```
[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/utils/utils.js:17]: I thought it would be "uglier" to add the wcpayPaymentRequestParams global variable in React everywhere, so I added this "getter function", so that we know where it's coming from. It's supposed to look similar to `getStripeServerData` from the Stripe implementation.

Do you think it's better to use `wcpayPaymentRequestParams?.key` along with `/* global wcpayPaymentRequestParams */` at the header instead? I'm happy to change.

File: client/payment-request/utils/normalize.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ * @param {Object} paymentData Stripe's order object.
+ *
+ * @return {Object} Order object in the format WooCommerce expects.
+ */
+export const normalizeOrderData = ( paymentData ) => {
+	const name = paymentData.paymentMethod.billing_details.name;
+	const email = paymentData.paymentMethod.billing_details.email;
+	const phone = paymentData.paymentMethod.billing_details.phone;
+	const billing = paymentData.paymentMethod.billing_details.address;
+	const shipping = paymentData.shippingAddress ?? {};
```
[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/utils/normalize.js:33]: It makes sense to play safe here. It even fixes this problem I was about to fix in a separate PR: #1677

Thanks!

**Comment by ricardo** [comment]
> Just noticed that Google Pay does not have a light-outline theme; only dark and light:

Great catch! I changed the behavior a little bit, after looking at how Stripe renders the actual Google Pay button. The `light` version does not have a `box-shadow`.

I had to add a small vertical margin so that the ~~outline~~ shadow doesn't get cut by `overflow: hidden` in the blocks layout.

Fixed in [ffb0de4](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/ffb0de4a17cca2ba41cea09e35b9a0404cad50f4)

> It seems the aforementioned issue with failing 3DS verification through a PRB might be an issue that existed before this PR, since the product page shows the exact same result:

The issue was actually generated from this PR 😬 

Fixed it in [eb615a7](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/eb615a77c6b859d26e0ba97c78b7d4fba23e1f29).

File: client/payment-request/event-handlers.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (reykjalin) [reply client/payment-request/event-handlers.js:58]: > it can return a confirmation Promise, which then resolves to a redirect URL.

Ahhh, in that case I think we should keep the `true === confirmation` check. And it also makes sense to keep the `confirmation` variable name.

>I didn't understand the reasoning behind calling it the opposite name. Can you explain?

No 😅 

I don't know why it's called the opposite. Maybe the blame will reveal who worked on this last? I doubt they remember the reasoning though 😬

File: client/payment-request/utils/utils.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (reykjalin) [reply client/payment-request/utils/utils.js:17]: Ahhh, that makes sense actually; wrapping the global variable like that. I didn't connect the dots there.

I think I'd prefer to keep the function, provided that we make it clearer in the function body that `wcpayPaymentRequestParams` is a global? Possibly with a `global wcpayPaymentRequestParams` at the top of the function — if that doesn't break anything — or a comment?

[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/utils/utils.js:17]: There's a `/* global wcpayPaymentRequestParams */` comment at the top, where we usually place this sort of comment. Marking as resolved.

**Comment by reykjalin** [comment]
> > This is sort-of unrelated to this PR, but I'm reporting it here anyway; @ricardo please let me know if I should move this to a different issue/PR
> 
> Please move to a new issue 👍

Issue opened in https://github.com/Automattic/woocommerce-payments/issues/1706

File: client/payment-request/style.scss [hunk 5]
```diff
@@ -11,26 +11,40 @@
...
 		&:active {
 			background-color: #fff;
 		}
 		&:hover {
 			background-color: #f8f8f8;
 		}
 		&:focus {
-			box-shadow: #e8e8e8 0 1px 1px 0, #e8e8e8 0 1px 3px;
 			outline: 0;
 		}
```
Comment (reykjalin) [inline client/payment-request/style.scss:49]: We need to include the `box-shadow` here to indicate whether the element is focused when (for example) navigating via the <kbd>tab</kbd> key.

Personally, I think we should allow the purple outline set by WC Core, since that's really universal throughout WC pages; I don't see why we would remove that. Maybe @LevinMedia has some thoughts on that, design-wise?
The `box-shadow` on its own is **very difficult** to notice.

We should also make sure this is the same across both Stripe and WCPay, so what we decide to do here, I will also do in Stripe. For now I just added the `box-shadow` back in Stripe.

File: client/payment-request/style.scss [hunk 6]
```diff
@@ -46,13 +60,13 @@
 			background-color: #3c4043;
 		}
 		&:focus {
-			box-shadow: #5f6368 0 1px 1px 0, #5f6368 0 1px 3px;
 			outline: 0;
 		}
```
Comment (reykjalin) [inline client/payment-request/style.scss:64]: Same as above ^, we need to include the `box-shadow` at the very least, and maybe allow the outline too.

**Review by reykjalin — CHANGES_REQUESTED** [review]
Just 2 minor comments on the updated PRB styling, but otherwise this looks good!

All the issues I encountered during my previous review have been fixed! Awesome work! 👏 

Let's get that CSS fixed (possibly with the help of @LevinMedia) and then get this shipped 🎉

File: client/payment-request/style.scss (hunk 5 above)
Comment (ricardo) [reply client/payment-request/style.scss:49]: Good point about the accessibility issue.

I removed the `box-shadow` because the Google Pay button that Stripe renders (when you have the wallet available) doesn't have a `box-shadow`.

Apparently, it uses the same styling as `:hover`, so perhaps the right thing to do is to use:

```scss
&:hover, &:focus {
	background-color: #f8f8f8;
}
```

image

Comment (ricardo) [reply client/payment-request/style.scss:49]: We can remove the `box-shadow` styling from `:focus` in the `light-outline` theme too. Stripe adds the same styling as in `:hover`.

image

File: client/payment-request/style.scss (hunk 6 above)
Comment (ricardo) [reply client/payment-request/style.scss:64]: Stripe also removes the outline property. I guess the best here is to be compliant with Stripe's standards. Using the same `background-color` as in `:hover` fixes this.

image

File: client/payment-request/style.scss (hunk 5 above)
Comment (reykjalin) [reply client/payment-request/style.scss:51]: The Stripe plugin version of this is not the best version of this either IMO, so I'd really like to get @LevinMedia's 👀 on this 🙂 

I'll ping him in Slack 👍

**Review by reykjalin — APPROVED** [review]
This tests and works really well!

At least the most common use cases; I know you've been working on some edge cases, but the only one I (know of and) tested was where removing a product and leaving only a virtual product in the cart wasn't removing shipping options from the PRB.

I like the updated CSS for the PRBs too; you're right setting the background color to highlight it works well. I still think we should allow the purple outline, but that's not really up to me (cc @LevinMedia) 😛 

The code looks great too! I have one last minor suggestion, but that's not a blocker for merging 🙂 

I think this looks great, and it's time to ship! 🚀 
If there are more edge cases you want me to test, please let me know and I'll run through  them ☺️

File: client/payment-request/blocks/use-initialization.js [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+		window.location = redirectUrl;
+	};
+
+	// Whenever paymentRequest changes, hook in event listeners.
+	useEffect( () => {
+		const noop = { removeAllListeners: () => void null };
+		let shippingAddressChangeEvent = noop,
+			shippingOptionChangeEvent = noop,
+			paymentMethodChangeEvent = noop,
+			cancelChangeEvent = noop;
```
[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/blocks/use-initialization.js]: @reykjalin Fixed in [`bef480e`](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/bef480e613ee5603752fb9f07c6b26c486437c30).

I went a little further and removed the individual `removeAllListeners()` by using `paymentRequest.removeAllListeners()` at the top.

To test, I added a `console.log` to the event handlers to see if they are't being triggered more than once, and this seems to work well.

[the remark this answers is not in the document]
Comment (reykjalin) [reply client/payment-request/blocks/use-initialization.js]: > I went a little further and removed the individual `removeAllListeners()` by using `paymentRequest.removeAllListeners()` at the top.

I don't know how I feel about that tbh. The function returned from `useEffect` is intened to [clean up resources before the component leaves the screen](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect).

I don't think there's anything wrong with the approach in this particular case, but it's "bad form" so to speak. I think it'd be better to return a function that calls `paymentRequest.removeAllListeners()` 🤔 :

```js
// add handlers
const shippingAddressChangeHandler = paymentRequest?.on( '', () => {} );
const anotherHandler = paymentRequest?.on( // ...

return () => {
    paymentRequest?.removeAllListeners();
}
```

It feels a bit better to use the cleanup function for this purpose, but I'm not _strictly opposed_ to the current approach. I would say I'd strongly recommend going with the cleanup function calling `removeAllListeners()` 🙂

[the remark this answers is not in the document]
Comment (ricardo) [reply client/payment-request/blocks/use-initialization.js]: @reykjalin Fixed in [`5b2980c`](https://github.com/Automattic/woocommerce-payments/pull/1529/commits/5b2980c33074965c9da6b9f48535aeede1c5ab6a), let me know how that looks 🙂

[the remark this answers is not in the document]
Comment (reykjalin) [reply client/payment-request/blocks/use-initialization.js]: Looks awesome! 🎉

**Comment by LevinMedia** [comment]
@reykjalin @ricardo - let's just use the styling that stripe provides for now - I'm hesitant to use a specific color outline for focus since it may not always work with a theme's styles 👍

File: client/payment-request/event-handlers.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	event.updateWith( {
+		status: response.result,
+		shippingOptions: response.shipping_options,
+		total: response.total,
+		displayItems: response.displayItems,
+	} );
+};
+
+export const shippingOptionChangeHandler = async ( api, event ) => {
+	const response = await api.paymentRequestUpdateShippingDetails( event );
```
Comment (asumaran) [inline client/payment-request/event-handlers.js:25]: @ricardo it should've been `event.shippingOption`. I'm submitting a PR fixing it.
